### PR TITLE
added preliminary timetable parsing capabalities. Successfully scrapes a...

### PR DESCRIPTION
**Pull request by shaftoes**
...ll departments except for computer science, online courses, and first year seminars. time slots are not properly inserted.

### CODE REVIEW

**Comment by shaftoes** [comment]
Sorry it's late, I pushed it last night but forgot to pull!

**Comment by david-yz-liu** [comment]
@shaftoes Since I merged web-request into master, you can start making PRs directly to master, actually.

File: hs/WebParsing/TimeTableParser.hs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+expandNote row =  if ( (T.take 4 (head row)) == "NOTE" )
+                  then ["", "", ""] ++ (tail row)
+                  else row
+
+{----------------------------------------------------------------------------------------
+takes in a department page name, extracts the html table, partitions into a list of all
+information related to a single course, and inserts the resulting tutorials and lectures
+into the database.
+----------------------------------------------------------------------------------------}
+getDeptTimetable :: String -> IO()
```
Comment (david-yz-liu) [inline hs/WebParsing/TimeTableParser.hs]: Space after IO

File: hs/WebParsing/TimeTableParser.hs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+getDeptTimetable url = do
+  rsp <- simpleHTTP (getRequest $ timetableUrl ++ url)
+  body <- getResponseBody rsp
+  let rawSoup = map cleanTag (parseTags (T.pack body))
+      table = dropAround  (tagOpen (\x -> or [(x =="table"), (x=="TABLE")]) (\x -> True))
```
Comment (david-yz-liu) [inline hs/WebParsing/TimeTableParser.hs]: If there's only two clauses, you should use `(||)`

File: hs/WebParsing/TimeTableParser.hs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+----------------------------------------------------------------------------------------}
+getDeptTimetable :: String -> IO()
+getDeptTimetable url = do
+  rsp <- simpleHTTP (getRequest $ timetableUrl ++ url)
+  body <- getResponseBody rsp
+  let rawSoup = map cleanTag (parseTags (T.pack body))
+      table = dropAround  (tagOpen (\x -> or [(x =="table"), (x=="TABLE")]) (\x -> True)) 
+                          (tagClose (\x -> or [(x=="table"), (x=="TABLE")])) 
+                          rawSoup
+      cells = filter (\x ->  and [(x /= []),(length x) > 4, notCancelled (x !! 4)])  (toCells table)
```
Comment (david-yz-liu) [inline hs/WebParsing/TimeTableParser.hs]: Don't need brackets around `length x`

File: hs/WebParsing/TimeTableParser.hs [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+      expandedNote = map expandNote cells
+      courseCells = partitions (\row -> ((head row) /= "")) expandedNote
```
Comment (david-yz-liu) [inline hs/WebParsing/TimeTableParser.hs]: Inner parentheses not necessary

File: hs/WebParsing/TimeTableParser.hs [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    notCancelled "" = True
+    notCancelled str = (T.head str) /= 'C'
+
+{----------------------------------------------------------------------------------------
+partitions the html table into a 2d list of cells. Does not account for cells that take
+up more than one row or column.
+----------------------------------------------------------------------------------------}
+toCells :: [Tag T.Text] -> [[T.Text]]
+toCells tags = 
+  let row = partitions (tagOpen (=="tr") (\x -> True)) tags
```
Comment (david-yz-liu) [inline hs/WebParsing/TimeTableParser.hs]: You should do `\_ -> True`

File: hs/WebParsing/TimeTableParser.hs [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+makeTutorial slot (Just tutorial) =
+  tutorial {timeStr = T.append (timeStr tutorial) (slotTime_str slot)}
+
+{----------------------------------------------------------------------------------------
+makes a lecture from the information currently residing in the courseSlot
+----------------------------------------------------------------------------------------}
+makeLecture :: CourseSlot -> Maybe Lecture -> Lecture
+makeLecture slot Nothing = 
+ Lecture { extra = (slotExtra slot),
+              section = (slotSection slot),
```
Comment (david-yz-liu) [inline hs/WebParsing/TimeTableParser.hs]: The indentation here is off

Comment (david-yz-liu) [inline hs/WebParsing/TimeTableParser.hs]: Also, I don't think any of these parentheses are necessary

File: hs/WebParsing/TimeTableParser.hs [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            }
+makeLecture slot (Just lec) = lec {time_str = T.append (time_str lec) (slotTime_str slot)}
+
+{----------------------------------------------------------------------------------------
+updates the courseSlot record with information from a given row
+----------------------------------------------------------------------------------------}
+updateSlot :: [T.Text] -> CourseSlot -> CourseSlot
+updateSlot row slot =
+  let partialUpdate = slot {
+            slotTime_str = T.takeWhile (/= ' ') (row !! 5),
```
Comment (david-yz-liu) [inline hs/WebParsing/TimeTableParser.hs]: Indentation

File: hs/WebParsing/TimeTableParser.hs [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+
+there are five cases, the last of which is useless and only in place for pattern complet-
+eness. 
+  1. no more information, with last block being a tutorial. 
+  2. no more information, with last block being a lecture.
+  3. moer information with last block being a lecture
+  4. more information with last block being a tutorial.
+within 3. and 4, if the current row is 
+----------------------------------------------------------------------------------------}
+updateSession :: [[T.Text]] -> CourseSlot -> Maybe Tutorial -> Maybe Lecture -> Session -> Session
```
Comment (david-yz-liu) [inline hs/WebParsing/TimeTableParser.hs]: Look into the `Either` type here instead of using two parameters for each of a tutorial and lecture

**Pull request closed without merging**